File: docsapp/com.ibm.concord.spreadsheet/src/com/ibm/concord/spreadsheet/document/model/impl/ReferenceList.java
package com.ibm.concord.spreadsheet.document.model.impl;

/**
 * Store all the formula references and name ranges
 */
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;
import java.util.Set;
import java.util.logging.Level;
import java.util.logging.Logger;

import com.ibm.concord.spreadsheet.common.ReferenceParser.ParsedRef;
import com.ibm.concord.spreadsheet.common.ReferenceParser.ParsedRefType;
import com.ibm.concord.spreadsheet.conversion.impl.ConversionUtil.RangeUsage;
import com.ibm.concord.spreadsheet.document.model.NotifyEvent.TYPE;
import com.ibm.concord.spreadsheet.document.model.util.ModelHelper;

public class ReferenceList extends RangeList<Integer>
{
  private static final Logger LOG = Logger.getLogger(ReferenceList.class.getName());
  
  private int maxRefCount = 0;
  
  public ReferenceList(Document doc)
  {
    super(doc);
  }

  /**
   * add reference into referencelist only when there is no such reference
   * otherwise just add ref count for it
   * precondition: the given reference must contain rangeId
   */
  public Reference addRange(Range<Integer> ref)
  {
    if(ref == null)
      return null;
    if(ref.getId() == IDManager.INVALID)
    {
      LOG.log(Level.WARNING, "the given ref must has range id when addRange");
      return null;
    }
    Reference r = null;
    RangesStruct sheet = this.getRangesStruct(ref.getSheetId(), false);
    if(sheet != null)
    {
      r = (Reference)sheet.getRangeMap().get(ref.getId());
    }
    
    if(r == null)
      r = (Reference)super.addRange(ref);
    
    r.addCount();
    
    return r;
    
  }
  
  public void deleteRange(Range<Integer> ref)
  {
    this.deleteRange((Reference)ref, false);
  }
  
  /**
   * delete reference ref count from reference list
   * if ref count is less than 1, remove from reference list
   * except name range
   * @param ref will be deleted reference
   * @param bDeleteUsage  delete reference no matter the refcount
   */
  public void deleteRange(Reference ref, boolean bDeleteUsage)
  {
    if(ref != null)
    {
      ref.deleteCount();
      
      if((ref.getCount() <= 0 && (ref.getUsage() != RangeUsage.NAMES) ||  bDeleteUsage))
      {
        super.deleteRange(ref);
      }
    }
  }

  /**
   * return the reference with the given ref address and ref type
   * this method will not add reference into reference lsit
   * @param ref
   * @param sheetName
   * @param type
   * @param bCreateRefId
   * @return
   */
  public Reference getRefByAddress(ParsedRef ref, String sheetName, int type, boolean bCreateRefId)
  {
    Reference reference = find(ref, sheetName, type);
    if(reference != null)
    {
      if (bCreateRefId && (reference.id == IDManager.INVALID))
      {
        reference.setId(generateId());
        //if it does not contain id, it means it is a new reference,
        //and it want the id, it called by generateToken without push to cell's tokenList
        //TODO: more test
        super.addRange(reference);
      }
    }
    return reference;
  }

  private Reference find(ParsedRef ref, String sheetName, int type)
  {
    Reference reference = new Reference(doc, ref, sheetName, true);
    reference.setRefType(type, false);
    RangesStruct sheet = this.getRangesStruct(reference.sheetId, false);
    if(sheet != null)
    {
      //check if there are the same reference in the list
      Set<Reference> set;
      if(reference.startRowId != IDManager.INVALID && ref.type != ParsedRefType.COLUMN)
        set = (Set<Reference>)sheet.getRows().get(reference.startRowId);
      else
        set = (Set<Reference>)sheet.getCols().get(reference.startColId);
      if (set != null)
      {
        Iterator<Reference> iter = set.iterator();
        while(iter.hasNext())
        {
          Reference r = iter.next();
          if(r.equals(reference))
          {
            //TODO: when calculate enabled
            return r;
          }
        }
      }
    }
    return reference;
  }

  public int generateId()
  {
    return maxRefCount++;
  }

  //override
  protected void modifyRangesDataByUsage(int sheetId, int sIndex, int eIndex, TYPE type)
  {
    // NOTHING todo.. for reference list predelete, 
  }
  
  // TODO: recoversheetrefs?
  
  /**
   * Used for recover manager from the reference list in main document to backup all the reference and name range in sheet which has sheetId as key
   * @param sheetId the will be deleted sheet id
   * @param refList   the reference list of the main document
   */
  public void backup(int sheetId, RangeList<Integer> refList)
  {
    RangesStruct<Integer> struct = refList.map.get(sheetId);
    //only back up range list when main document has ranges in sheetId
    if(struct != null)
    {
      if(map.get(sheetId) != null)
      {
        LOG.log(Level.WARNING, "the map for sheet {0} should not exist in range list of recover manager", new ModelHelper.SerializableStringIdConvertor().toStringId(sheetId));
        return;
      }
      if(this.doc == refList.doc)
      {
        LOG.log(Level.WARNING, "when backup range list, doc of recover manager is the same with the main document");
        return;
      }
      map.put(sheetId, struct);
      //update usage map and nameRangeMap, because when serialize name range, it iterate the name range map
      RangeMap<Integer> ranges = struct.getRangeMap();
      Iterator<Integer> iter = ranges.keySet().iterator();
      while(iter.hasNext())
      {
        int id = iter.next();
        Reference r = (Reference)ranges.get(id);
        // should set parent document when serialize the recover document
        // when call range.getAddress(), the address should be #REF!,
        // but if set name range parent to recover doc, the address will not be #REF!
//        r.parent = doc;
      }
    }
  }
  
  /**
   * ReferenceList does not contain usageMap
   */
  public Map<RangeUsage, RangeMap<Integer>> getByUsageRangeMap()
  {
    return null;
  }
  /**
   * Used to recover all the reference and name range of specify sheet from recover manager to main document
   * @param sheetId the will be recovered sheet id
   * @param recoverRefList the reference list of recover document
   */
  public void recoverRanges4Sheet(int sheetId, ReferenceList recoverRefList)
  {
    Map<Integer, RangesStruct<Integer>> sheets = recoverRefList.getBySheetIdRangeMap();
    RangesStruct<Integer> struct = sheets.get(sheetId);
    if(struct == null)
      return;
    map.put(sheetId, struct);
    recoverRefList.map.remove(sheetId);
    RangeMap<Integer> ranges = struct.getRangeMap();
    Iterator<Integer> rangeIter = ranges.keySet().iterator();
    while(rangeIter.hasNext())
    {
      int id = rangeIter.next();
      Reference ref = (Reference)ranges.get(id);
      ref.parent = doc;
      RangeUsage usage = ref.getUsage();
      if(usage == RangeUsage.NAMES)
      {
        ParsedRef parsedRef = ref.getParsedRef();
        this.updateRangeByUsage(parsedRef, ref, usage);
      }else
        ref.recoverSheetName(); 
      
    }
  }
}

File: docs-web.resources/com.ibm.docs.web.resources/WebContent/js/wseditor/js/JsProxyModel/ReferenceList.js
dojo.provide('websheet.JsProxyModel.ReferenceList');
//not used for now
dojo.declare("websheet.JsProxyModel.ReferenceList", websheet.functions.ReferenceList, {
	_doc: null,	//websheet.JsProxyModel.Document
	_idManager: null, //websheet.JsProxyModel.IDManager
	_nameRefs: null,	//the list for name reference
	_refs: null,		// the list for formula reference
	constructor: function(d, idmanager)
	{
		this._doc = d;
		this._idManager = idmanager;
		this._nameRefs = {};
		this._refs = {};
	},
	
	getRefByAddress:function(address, sheetName, bCreate, refType, ignoreInvalid){
		//TODO: the same address are not share the same reference
		var ref = new websheet.JsProxyModel.Reference(this._doc, this._idManager);
		if(refType == undefined)
			refType = 0;
		
		ref._setRefType(refType,false);
		ref.parseAddress(address, sheetName);
		var name = ref._rangeInfo.sheetName;
		var list = this._refs[name];
		if(!list){
			list = {};
			this._refs[name] = list;
		}
		var address = ref.getAddress();
		var rl = list[address];
		if(rl != null){
			var length = rl.length;
			for(var i = 0; i < length; i++){
				var r = rl[i];
				if(r.getRefType() == refType)
					return r;
			}
		}
		return ref;
	},
	
	addRange: function(range){
		if(range ){
			if(range.getUsage() == websheet.Constant.RangeUsage.NAME)
				this._nameRefs[range.getRangeId()] = range;
			var name = range._rangeInfo.sheetName;
			var list = this._refs[name];
			if(!list){
				list = {};
				this._refs[name] = list;
			}
			var address = range.getAddress();
			var rl = list[address];
			if(!rl)
				list[address] = [range];
			else{
				var bFound = false;
				var length = rl.length;
				var refType = range.getRefType();
				for(var i = 0; i < length; i++){
					var r = rl[i];
					if(r.getRefType() == refType){
						bFound = true;
						break;
					}
				}
				if(!bFound)
					rl.push(range);
			}
			range.addCount();
		}
		//do nothing
	},
	
	deleteRange:function(range){
		if(!range)
			return;
		var count = range.deleteCount();
		//TODO necessary?
		if(count <= 0){
			range.resetCache();
			var name = range._rangeInfo.sheetName;
			var list = this._refs[name];
			if(list){
				var address = range.getAddress();
				var rl = list[address];
				var length = rl.length;
				var refType = range.getRefType();
				for(var i = 0; i < length; i++){
					var r = rl[i];
					if(r.getRefType() == refType){
						rl.splice(i,1);
						break;
					}
				}
				if(rl.length == 0)
					delete list[address];
			}
		}
	},
	
	getRange:function(){
		console.warn("getRange does not supported yet");
	},
	
	getRangeByUsage:function(rangeId, usage){
		if(usage == websheet.Constant.RangeUsage.NAME && rangeId){
			var range = this._nameRefs[rangeId.toLowerCase()];
			return range;
		}else
			console.warn("getRangeByUsage does not supported yet");
	},
	
	addNonNameCell:function(cell){
		//Do nothing
	},
	
	decompose:function() {
		this._nameRefs = null;
		this._refs = null;
		this._doc = null;
		this._idManager = null;
	}

});
